commit 81e0ce79f2919dbd5f025894d29aa806af8695c7
Author: Bjørn Mork <bjorn@mork.no>
Date:   Thu Dec 3 19:24:20 2015 +0100

    usbnet: allow mini-drivers to consume L2 headers
    
    Assume the minidriver has taken care of all L2 header parsing
    if it sets skb->protocol.  This allows the minidriver to
    support non-ethernet L2 headers, and even operate without
    any L2 header at all.
    
    Signed-off-by: Bjørn Mork <bjorn@mork.no>
    Acked-by: Oliver Neukum <oneukum@suse.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

diff --git a/drivers/net/usb/usbnet.c b/drivers/net/usb/usbnet.c
index 0744bf2..0b0ba7e 100644
--- a/drivers/net/usb/usbnet.c
+++ b/drivers/net/usb/usbnet.c
@@ -324,7 +324,10 @@ void usbnet_skb_return (struct usbnet *dev, struct sk_buff *skb)
 		return;
 	}
 
-	skb->protocol = eth_type_trans (skb, dev->net);
+	/* only update if unset to allow minidriver rx_fixup override */
+	if (skb->protocol == 0)
+		skb->protocol = eth_type_trans (skb, dev->net);
+
 	dev->net->stats.rx_packets++;
 	dev->net->stats.rx_bytes += skb->len;

